store/postgres/src/metadata.rs

```diff
@@ -751,6 +753,62 @@ pub fn reassign_subgraph(
     }
 }
 
+// Does nothing if the error already exists. Returns the error id.
+fn insert_subgraph_error(conn: &PgConnection, error: SubgraphError) -> anyhow::Result<String> {
+    use subgraph_error as e;
+
+    let error_id = hex::encode(&stable_hash::utils::stable_hash::<SetHasher, _>(&error));
+    let SubgraphError {
+        subgraph_id,
+        message,
+        handler,
+        block_ptr,
+        deterministic,
+    } = error;
+
+    let block_num = match block_ptr {
+        None => {
+            assert_eq!(deterministic, false);
+            crate::block_range::BLOCK_UNVERSIONED
+        }
+        Some(block) => crate::block_range::block_number(&block),
+    };
+
+    insert_into(e::table)
+        .values((
+            e::id.eq(&error_id),
+            e::subgraph_id.eq(subgraph_id.as_str()),
+            e::message.eq(message),
+            e::handler.eq(handler),
+            e::deterministic.eq(deterministic),
+            e::block_hash.eq(block_ptr.as_ref().map(|ptr| ptr.hash.as_bytes())),
+            e::block_number.eq(sql(&format!("{}::numeric", block_num))),
+            e::block_range.eq((Bound::Included(block_num), Bound::Unbounded)),
+        ))
+        .on_conflict_do_nothing()
+        .execute(conn)?;
```

Comment:
You also need to clamp the range of precious errors, otherwise, if you insert an error with range `[1,)`, you will never be able to insert another error for that subgraph.

And since `block_number = lower(block_range)`, do we even need that column?

Reply:
Since `on_conflict_do_nothing()` can mask problems silently, we need tests that check the various combinations of storing errors and that they work; at a minimum, it should check
* storing the same error twice stores only one error
* storing the same error at two different blocks stores two errors
* reverting correctly reverts errors

And it would be really nice if errors had a `created_at` column to make it easier to find new errors.

Reply:
Dropped `block_number`, added `created_at`, added some tests that the conflict do nothing works.